File: tray-app-win/MailMCP/IPC/Paths.cs
namespace MailMCP.IPC;

/// <summary>
/// Mirrors the Windows branch of <c>mail-mcp-core::paths::Paths::default_for_user</c>.
/// Override every directory under a single root via the <c>MAIL_MCP_ROOT</c>
/// env var (matches the daemon CLI flag).
/// </summary>
public sealed record MailMCPPaths(
    string DataDir,
    string LogsDir,
    string CacheDir,
    string RuntimeDir,
    string IpcPipe)
{
    public string EndpointJson => Path.Combine(DataDir, "endpoint.json");
    public string PidFile => Path.Combine(RuntimeDir, "daemon.pid");
    public string StateDb => Path.Combine(DataDir, "state.db");

    public static MailMCPPaths DefaultForUser()
    {
        var root = Environment.GetEnvironmentVariable("MAIL_MCP_ROOT");
        if (!string.IsNullOrEmpty(root))
        {
            return new MailMCPPaths(
                DataDir: Path.Combine(root, "data"),
                LogsDir: Path.Combine(root, "logs"),
                CacheDir: Path.Combine(root, "cache"),
                RuntimeDir: Path.Combine(root, "run"),
                IpcPipe: PipeName());
        }
        var local = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
        var appDir = Path.Combine(local, "mail-mcp");
        return new MailMCPPaths(
            DataDir: appDir,
            LogsDir: Path.Combine(appDir, "logs"),
            CacheDir: Path.Combine(appDir, "cache"),
            RuntimeDir: Path.Combine(appDir, "run"),
            IpcPipe: PipeName());
    }

    /// <summary>
    /// The named-pipe address: <c>\\.\pipe\mail-mcp-{USERNAME}</c>. Per-user
    /// scoping keeps separate user accounts isolated. Username is enough;
    /// SIDs require P/Invoke and aren't worth the complexity here.
    /// </summary>
    private static string PipeName()
    {
        var user = string.IsNullOrEmpty(Environment.UserName) ? "default" : Environment.UserName;
        return $@"\\.\pipe\mail-mcp-{user}";
    }
}

File: tray-app/MailMCP/IPC/Paths.swift
import Foundation

/// Mirrors the macOS branch of `mail-mcp-core::paths::Paths::default_for_user`.
/// Override via `MAIL_MCP_ROOT` env var (matches the daemon CLI flag).
public struct MailMCPPaths {
    public let dataDir: URL
    public let logsDir: URL
    public let cacheDir: URL
    public let runtimeDir: URL

    public var endpointJSON: URL { dataDir.appendingPathComponent("endpoint.json") }
    public var ipcSocket: URL    { runtimeDir.appendingPathComponent("ipc.sock") }
    public var pidFile: URL      { runtimeDir.appendingPathComponent("daemon.pid") }
    public var stateDB: URL      { dataDir.appendingPathComponent("state.db") }

    public static func defaultForUser() -> MailMCPPaths {
        if let root = ProcessInfo.processInfo.environment["MAIL_MCP_ROOT"] {
            return withRoot(URL(fileURLWithPath: root))
        }
        let home = FileManager.default.homeDirectoryForCurrentUser
        let data  = home.appendingPathComponent("Library/Application Support/mail-mcp")
        let logs  = home.appendingPathComponent("Library/Logs/mail-mcp")
        let cache = home.appendingPathComponent("Library/Caches/mail-mcp")
        let tmp = ProcessInfo.processInfo.environment["TMPDIR"]
            .map { URL(fileURLWithPath: $0) } ?? URL(fileURLWithPath: "/tmp")
        let uid = getuid()
        let runtime = tmp.appendingPathComponent("mail-mcp-\(uid)")
        return MailMCPPaths(dataDir: data, logsDir: logs, cacheDir: cache, runtimeDir: runtime)
    }

    public static func withRoot(_ root: URL) -> MailMCPPaths {
        MailMCPPaths(
            dataDir: root.appendingPathComponent("data"),
            logsDir: root.appendingPathComponent("logs"),
            cacheDir: root.appendingPathComponent("cache"),
            runtimeDir: root.appendingPathComponent("run")
        )
    }
}
